README.rst
=================
 Printable Mazes
=================

This is akaihola's fork of David Bau's `original pymaze.py`_ script, which is
*"a free PDF maze generator that can create mazes of various sizes and
complexity, including pretty diabolical mazes that include 3-d
crossings..."*. `Dave's site`_ also includes a web form for generating maze
PDFs interactively using the code as a CGI script.


Requirements
============

The maze generator requires Python_ and the ReportLab_ library. On Debian-based
systems (e.g. Ubuntu) they must be installed from the command line::

    sudo apt-get install python-reportlab


Command-line arguments
======================

When run on the command line, the script accepts the following options:

  * ``-w <NUM>``: width of the paper (default: width of a letter page)
  * ``-h <NUM>``: height of the paper (default: height of a letter page)
  * ``-m <NUM>``: page margin (default: 36 points)
  * ``-p <PAGESIZE>``: page size (e.g. a4 or letter)
  * ``--cell=<NUM>``: cell size (default: 24)
  * ``--tube=<NUM>``: tube width (default: 0.7)
  * ``--wall=<NUM>``: wall thickness (default: 0.3)
  * ``--curve=<0/1>``: rounded corners (default: 1=yes)
  * ``--cross=<0/1>``: generate multi-level crossings (default: 1=yes)

Dimensions are expressed in *points* (1/72ths of an inch) except for tube width
and wall thickness, which are ratios relative to the cell size.

Example
=======

::

    python pymaze.py -p a4 --cross=0 --cell=48 >maze.pdf

.. _`original pymaze.py`: http://davidbau.com/downloads/pymaze.py
.. _`Dave's site`: http://davidbau.com/archives/2006/10/10/printable_mazes.html
.. _Python: http://python.org/
.. _ReportLab: http://www.reportlab.com/ 


CHANGELOG.rst
======================
 pypdfmaze change log
======================


2006-10-10 (David Bau)
======================

* published the `original script`_ as a `blog post`_


2011-10-13 (akaihola)
=====================

* added the ``-p`` page size argument
* created a `GitHub repository`_
* added a readme file
* added a change log


.. _`original script`: http://davidbau.com/downloads/pymaze.py
.. _`blog post`: http://davidbau.com/archives/2006/10/10/printable_mazes.html
.. _`GitHub repository`: https://github.com/akaihola/pypdfmaze


pymaze.py
#!/usr/bin/env python2.6
#
# Maze generator in 130 lines (non-blank/comment) of Python
# Copyright (c) 2008 David Bau.  All rights reserved.
# Modifications (c) 2011 Antti Kaihola: https://github.com/akaihola/pypdfmaze
#
# Can be used as either a command-line tool or as a cgi script
# Numeric arguments are w, h, m, cell, tube, wall, curve, and cross.
# String argument -p specifies the page size, e.g. a4 or letter.

import cgitb
cgitb.enable()
import sys
import os
import random
import getopt
import cgi
import user
from reportlab.lib import pagesizes
from reportlab.pdfgen.canvas import Canvas
from StringIO import StringIO

def main():
  opts = {}
  for k, v in getopt.getopt(sys.argv[1:], 'w:h:m:p:',
              ['cell=', 'tube=', 'wall=', 'curve=', 'cross='])[0]:
    optname = k.strip('-')
    if optname != 'p':
      # the -p option is a string (e.g. "-p a4" or "-p letter"),
      # other options are int/float
      v = num(v)
    opts[optname] = v
  if 'p' in opts:
    # the -p option (e.g. 'a4' or 'letter') sets width and height, but they may
    # be overridden with -w and -h
    width, height = getattr(pagesizes, opts.pop('p').upper())
    opts.setdefault('w', width)
    opts.setdefault('h', height)
  drawmaze(sys.stdout, **opts)

def cgimain():
  import cgitb; cgitb.enable()
  pdfout = StringIO()
  print >> pdfout, "Content-Type: application/pdf"
  print >> pdfout, "Cache-Control: no-cache"
  print >> pdfout, "Expires: Sat, 01 Jan 2000 00:00:00 GMT"
  print >> pdfout
  opts = {}
  fieldStorage = cgi.FieldStorage()
  for key in fieldStorage.keys():
    opts[key] = num(fieldStorage[key].value)
  drawmaze(pdfout, **opts)
  print >> sys.stdout, pdfout.getvalue()

def drawmaze(file, w = 8.5 * 72, h = 11.0 * 72, m = 36,
             cell = 24, tube = 0.7, wall = 0.3, curve = 1, cross = 1,
             count = 1):
  cellw, cellh = [int((x - 2 * m) / cell) for x in (w, h)]
  mx, my = (w - cellw * cell) / 2.0, (h - cellh * cell) / 2.0
  if cellw <= 0 or cellh <= 0: raise RuntimeError('Bad maze dimensions')
  c = Canvas(file)
  c.setTitle('Maze')
  c.setSubject('%s by %s %s crossings' %
               (cellw, cellh, ('without', 'with')[cross]))
  c.setAuthor("Dave's Maze Maker")
  c.setPageSize((w, h))
  for n in xrange(count):
    maze = [0] * cellw * cellh
    fillmaze(maze, cellw, cellh, cross)
    maze[0] |= 4
    maze[cellw * cellh - 1] |= 1
    for x in xrange(cellw):
      for y in xrange(cellh):
        drawcell(c, maze[x + cellw * y],
                 x * cell + mx, y * cell + my, cell, tube, wall, curve)
    c.showPage()
  c.save()

def drawcell(canvas, bits, x, y, size, shape, wall, curve):
  a, b, lw = size * 0.5, size * 0.5 * shape, size * 0.5 * wall
  over = (0 != (bits & 16))
  canvas.saveState()
  canvas.setLineWidth(lw)
  canvas.setLineCap(curve and 1 or 2)
  canvas.translate(x + a, y + a)
  gaps = rgaps(bits)
  if gaps:
    canvas.rotate((gaps[0] + 1) * 90)
  for gap in gaps[1:]:
    p = canvas.beginPath()
    p.moveTo(b, -a)
    if gap == 0:
      if curve: p.arcTo(b, b - a - a, a + a - b, -b, 180, -90)
      else: p.lineTo(b, -b); p.lineTo(a, -b)
    elif gap == 1:
      p.lineTo(b, a)
      if over: p.moveTo(a, -b); p.lineTo(b, -b); p.moveTo(a, b); p.lineTo(b, b)
    elif gap == 2:
      if curve: p.arcTo(-b - a - a, -b - a - a, b, b, 0, 90)
      else: p.lineTo(b, b); p.lineTo(-a, b)
    elif gap == 3:
      if curve: p.arcTo(-b, -b, b, b, 0, 180)
      else: p.lineTo(b, b); p.lineTo(-b, b)
      p.lineTo(-b, -a)
    canvas.drawPath(p)
    canvas.rotate((gap + 1) * 90)
  canvas.restoreState()

def rgaps(bits):
  result, count = [], 0
  for bit in xrange(4):
    if not (bits & 1 << bit): count += 1
    else: result.append(count); count = 0
  if result:
    result.append(count + result[0])
  return result

def fillmaze(maze, w, h, over):
  path = random.sample([pos for pos in xrange(w * h) if maze[pos] == 0], 1)
  nextdir = -1
  while path:
    pos = path[-1]
    choices = [d for d in xrange(4) if canbuild(maze, w, h, pos, d, over)]
    if nextdir in choices and random.randint(0, 1) == 0: pass
    elif choices: nextdir = random.sample(choices, 1)[0]
    else: path.pop(); continue
    path.append(dobuild(maze, w, h, pos, nextdir, over))

def atedge(w, h, pos, dir):
  if dir == 0: return pos % w == w - 1
  if dir == 1: return pos >= w * (h - 1)
  if dir == 2: return pos % w == 0
  if dir == 3: return pos < w

def cancross(maze, pos, dir):
  return maze[pos] & 15 == (10, 5, 10, 5)[dir]

def canbuild(maze, w, h, pos, dir, over):
  if dir < 0: return False
  offset = (1, w, -1, -w)[dir]
  while not atedge(w, h, pos, dir):
    pos += offset
    if maze[pos] == 0: return True
    if not over or not cancross(maze, pos, dir): return False
  return False

def dobuild(maze, w, h, pos, d, over):
  offset = (1, w, -1, -w)[d]
  maze[pos] |= 1 << d
  pos += offset
  while maze[pos] != 0:
    maze[pos] ^= random.choice((0, 15))
    maze[pos] |= 16
    pos += offset
  maze[pos] |= 1 << (d ^ 2)
  return pos

def num(s):
  if s.find('.') < 0: return int(s)
  else: return float(s)

if __name__ == "__main__":
  if os.getenv('GATEWAY_INTERFACE') is not None: sys.exit(cgimain())
  else: sys.exit(main())

